pkg/manifests/manifests.go

```diff
@@ -3897,22 +3904,24 @@ func (f *Factory) TelemeterClientServiceAccount() (*v1.ServiceAccount, error) {
 }
 
 // TelemeterClientSecret generates a new Secret for Telemeter client.
-func (f *Factory) TelemeterClientSecret() (*v1.Secret, error) {
+func (f *Factory) TelemeterClientSecret(secretToken string) (*v1.Secret, error) {
 	s, err := f.NewSecret(f.assets.MustNewAssetReader(TelemeterClientSecret))
 	if err != nil {
 		return nil, err
 	}
 
-	salt, err := GeneratePassword(32)
-	if err != nil {
-		return nil, fmt.Errorf("failed to generate Telemeter client salt: %v", err)
-	}
-	s.Data["salt"] = []byte(salt)
-
 	if f.config.ClusterMonitoringConfiguration.TelemeterClientConfig.Token != "" {
 		s.Data["token"] = []byte(f.config.ClusterMonitoringConfiguration.TelemeterClientConfig.Token)
 	}
 
+	if f.config.ClusterMonitoringConfiguration.TelemeterClientConfig.Token != secretToken {
+		salt, err := GeneratePassword(32)
+		if err != nil {
+			return nil, fmt.Errorf("failed to generate Telemeter client salt: %v", err)
+		}
+		s.Data["salt"] = []byte(salt)
```

Comment:
This only adds salt to the secret if the token has changed. In the [deployment of telemeter client](https://github.com/openshift/telemeter/blob/master/manifests/client/deployment.yaml) the secret is mounted and passed as `--anonymize-salt-file=/etc/telemeter/salt`, and I understand this would lead to a file not found error [here](https://github.com/openshift/telemeter/blob/master/pkg/forwarder/forwarder.go#L126). The Pod could be restarted due to some crash or reschedule, so I think the salt should always be defined in the secret. For example we could modify `(c *Client) GetTelemeterSecretToken` to return both the current token and the current salt and pass them both to this method

Reply:
So I've changed this slightly so that the salt is generated if the token changes or if the salt gets erased since a customer might do it by accident.

Reply:
But if the token has not changed and the salt exists, then `s.Data` will not contain the salt. Then `func (t *TelemeterClientTask) create` will do `err = t.client.CreateOrUpdateSecret(ctx, s)` which will replace the current secret with `s` after doing `mergeMetadata` with the existing secret which takes care of `ObjectMeta` but not the `Data` field. As a result, the secret in the k8s API won't have an entry `Data["salt"]`, which will be a problem if the telemeter pod is restarted. 
As seen in `vendor/k8s.io/client-go/kubernetes/typed/core/v1/secret.go` we have `func (c *secrets) Update` using `c.client.Put()` that is the [API for replacing a Secret object](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.22/#replace-secret-v1-core), this is not updating selected fields as with PATCH

Reply:
You are 💯 correct, I'm sorry I was overlooking this particularity 🤦